src/index.ts

```diff
@@ -74,253 +74,272 @@ const TYPED_MESSAGE_SCHEMA = {
             ),
-          ];
-        }
+          ),
+        ];
+      }
 
-        return [type, value];
-      };
+      return [type, value];
+    };
 
-      for (const field of types[primaryType]) {
-        const [type, value] = encodeField(
-          field.name,
-          field.type,
-          data[field.name],
-        );
-        encodedTypes.push(type);
-        encodedValues.push(value);
-      }
-    } else {
-      for (const field of types[primaryType]) {
-        let value = data[field.name];
-        if (value !== undefined) {
-          if (field.type === 'bytes') {
-            encodedTypes.push('bytes32');
-            value = ethUtil.keccak(value);
-            encodedValues.push(value);
-          } else if (field.type === 'string') {
-            encodedTypes.push('bytes32');
-            // convert string to buffer - prevents ethUtil from interpreting strings like '0xabcd' as hex
-            if (typeof value === 'string') {
-              value = Buffer.from(value, 'utf8');
-            }
-            value = ethUtil.keccak(value);
-            encodedValues.push(value);
-          } else if (types[field.type] !== undefined) {
-            encodedTypes.push('bytes32');
-            value = ethUtil.keccak(
-              this.encodeData(field.type, value, types, useV4),
-            );
-            encodedValues.push(value);
-          } else if (field.type.lastIndexOf(']') === field.type.length - 1) {
-            throw new Error(
-              'Arrays are unimplemented in encodeData; use V4 extension',
-            );
-          } else {
-            encodedTypes.push(field.type);
-            encodedValues.push(value);
+    for (const field of types[primaryType]) {
+      const [type, value] = encodeField(
+        field.name,
+        field.type,
+        data[field.name],
+      );
+      encodedTypes.push(type);
+      encodedValues.push(value);
+    }
+  } else {
+    for (const field of types[primaryType]) {
+      let value = data[field.name];
+      if (value !== undefined) {
+        if (field.type === 'bytes') {
+          encodedTypes.push('bytes32');
+          value = ethUtil.keccak(value);
+          encodedValues.push(value);
+        } else if (field.type === 'string') {
+          encodedTypes.push('bytes32');
+          // convert string to buffer - prevents ethUtil from interpreting strings like '0xabcd' as hex
+          if (typeof value === 'string') {
+            value = Buffer.from(value, 'utf8');
           }
+          value = ethUtil.keccak(value);
+          encodedValues.push(value);
+        } else if (types[field.type] !== undefined) {
+          encodedTypes.push('bytes32');
+          value = ethUtil.keccak(
+            encodeData(
+              field.type,
+              // TODO: Add validation to ensure this is a string-indexed
+              // object, so that this type cast can be removed
+              value as Record<string, unknown>,
```

Comment:
Here's the first type assertion that was required. This should be guaranteed to be a `Record<string, any>` for any properly formatted message, but we don't have sufficient type information at this point to guarantee that. We can add that later.